Allow scheduling system events persistently.

jankeromnes opened the pull request:
This is basically a re-work of https://github.com/JanitorTechnology/janitor/pull/140/ (sorry for the duplicate work!), which will allow us to schedule potentially long-term system events like "invalidate this token in 24 hours", "delete this container in 1 year", etc.

Here are the differences between this implementation and the "tasks" approach:

- I felt that using an `EventEmitter` was better-suited for registering + triggering callbacks on specific task/event types.
- I also regularly process upcoming tasks/events, but if they're due before the next loop, I use `setTimeout` to trigger them just in time.
- My approach doesn't need a unique task/event IDs.
- It supports multiple listeners for the same event type.
- It doesn't check for errors from the listeners (just if at least one listener was called for this event).

jankeromnes commented:
@nt1m and @bnjbvr could you please take a look at this approach? :)

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+// Copyright © 2018 Jan Keromnes. All rights reserved.
```
nit: All rights reserved is unnecessary in modern times, since this is [the default](https://opensource.stackexchange.com/questions/2121/mit-license-and-all-rights-reserved)

jankeromnes replied:
Yay for modern times! Removed the "All rights reserved." part, because it's both unnecessary and misleading.

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+  events.queue.forEach(event => { event.scheduled = null; });
+
+  // Start regularly scheduling events, once start-up is complete.
+  setTimeout(processQueue, schedulingStartDelay);
```
How did you determine that 1 minute was enough? Wouldn't it be clearer that there is an exported function `startScheduling` in this module that actually programs the first run, and when start-up has completed in the main module, it calls `startScheduling` instead?

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+  const now = Date.now();
+  if (event.time && event.time > now) {
+    // The event is not due yet, emit it later.
+    if (event.time < now + schedulingInterval && !event.scheduled) {
```
nit: use a non-overflowing comparison: `event.time - now < schedulingInterval`

bnjbvr replied:
(I also find the intent clearer by using this comparison)

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  // The event is due or past due, emit it now.
+  emitEvent(event);
```
`setTimeout(() => { emitEvent(event); }, 0);` ?

bnjbvr replied:
Also, this will reschedule events that have been scheduled but have no due time :)

jankeromnes replied:
All events have a due time. 😉

jankeromnes replied:
I don't see the point of `setTimeout(() => { emitEvent(event); }, 0);` as opposed to `emitEvent(event);` but fair enough, I'll do that. 😄

bnjbvr commented on api/admin-api.js:
```diff
@@ -13,6 +14,55 @@ const adminAPI = module.exports = selfapi({
+      return;
+    }
+
+    response.json(events.getQueue(), null, 2);
```
It's fine as it is now, but it could be an easy way to DDOS the server later (when there are many events), thus a paginated API would probably be better here.

jankeromnes replied:
Agreed, I do believe that we'll need to paginate our APIs at some point.

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+
+// Emit or schedule an event if it's due soon.
+function maybeSchedule (event) {
+  if (!event) {
```
I think you can simplify this whole function by having `if (!event || event.scheduled)` here, and then below you can coalesce the two `if`.

jankeromnes replied:
I tweaked the function following your suggestions. Hopefully you'll like how it looks now. 😄

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+  events.queue = events.queue.filter(event => !!event);
+
+  // Re-schedule any un-emitted events.
+  events.queue.forEach(event => { event.scheduled = null; });
```
Maybe assert that `event.emitted` is not set here? I think this can't happen (because the db saving happens at the end of the function that sets `emitted`, which  means that if the function gets interrupted, then the event will be scheduled twice)

jankeromnes replied:
Agreed, I don't think this can happen. Also, we don't use asserts in the code today, and I'm not really sure how to do that (e.g. What method should we use? Should we assert in production? Etc.)

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+function createEvent (type, time, payload = null) {
+  return {
+    type,
+    time,
```
I like all the attribute names you've used, but this one. How about `due_time` or `due_date` or `due_at`? (or in camelCase)

jankeromnes replied:
Fair enough, I went with `dueTime` to be very explicit.

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+    type,
+    time,
+    payload,
+    scheduled: null,
```
`scheduledAt`, to make it clear it's a date?

jankeromnes replied:
I went for `scheduledTime` to be consistent with `dueTime`.

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+    time,
+    payload,
+    scheduled: null,
+    emitted: null,
```
`emittedAt`?

jankeromnes replied:
Ok for `emittedTime`.

bnjbvr commented on tests/tests.js:
```diff
@@ -187,6 +187,62 @@ tests.push({
+        console.error(new Error('Received TestEvent too late at ' + delta));
+        return;
+      }
+      if (timeDelta <= tooEarlyDelta - expectedDelta) {
```
I think if `timeDelta === 1`, then it should be flagged as too early but this condition won't catch it. I think you want `Math.abs()` on both sides.

jankeromnes replied:
`tooEarlyDelta` and `expectedDelta` are constant values (currently `1000 * 2` and `1000 * 5`, respectively) so the right-hand value is `-3000` (it should always be negative).

`timeDelta` is `Date.now() - expectedTime`, so ideally it's close to `0` if the event is triggered on time, but it's negative when the event is received earlier and positive when it's received later.

So if the event is 1 millisecond late, I don't think that it should be flagged as too early. (Currently, we flag events "too early" when they are more than 3 seconds early, and we flag them "too late" when we receive them more than 3 seconds late.)

bnjbvr replied:
It's just conceptually weird that things that are after `expectedDelta` may not be always considered late. Reproducibility is hard, I admit it. Despite this, I think the concept of being late should be binary and not a blurry notion. That being said, it doesn't matter too much :)

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,137 @@
+  }
+
+  if (!events.queue) {
+    events.queue = [];
```
Can you comment, maybe in a block at the top of this file, what are the expected states of an event:

`in queue => scheduled => emitted`

jankeromnes replied:
Sure! Done. 😄

bnjbvr commented on tests/tests.js:
```diff
@@ -187,6 +187,62 @@ tests.push({
   }
 });
 
+tests.push({
```
Can you add a test for an event without a due date?

jankeromnes replied:
Events always have a due date. We could add a test for `exports.emit = function (eventType, payload = null) { exports.emitAtTime(eventType, Date.now(), payload); };` but given that it's just a simple wrapper, I don't see the value in testing it.

jankeromnes commented:
Thanks a lot for the review! 😄 I'll rebase these commits and address your comments/nits.

notriddle commented:
I just found out about [this "TaskList" library](https://github.com/IMA-WorldHealth/TaskList), if you're interested in outsourcing some of the complexity of this time ring implementation?

jankeromnes commented:
Addressed all nits and replied to all questions, except the nit about starting scheduling via an exported method instead of after an arbitrary 1min delay.

Also, I should look into notriddle's suggestion as an alternative to this, as it sounds promising.

jankeromnes commented:
> Also, I should look into notriddle's suggestion as an alternative to this, as it sounds promising.

Well, since this code is already written, maybe let's merge it as is without letting it bitrot any further.

TaskList looks great though, and its cron schedules look easy enough to persist and reload. It doesn't seem to allow irregular schedules (e.g. "every 30 hours", although that's not a strong requirement), and it has more than zero NPM dependencies (but not a huge problem either).

I should probably add a comment so that if we need more advanced scheduling functions, we switch to a real cron-based module instead of implementing and maintaining everything ourselves.

bnjbvr commented on lib/events.js:
```diff
@@ -0,0 +1,150 @@
+  // 1. Queued: `event` is added to `events.queue`
+  // 2. Scheduled: `event.scheduledTime` is set
+  // 3. Emitted: `event.emittedTime` is set, and `event` is moved from
+  //   `events.queue` to `events.history`
```
(I might be nitpicky here) Maybe this description is incomplete? It seems the original entry in the `queue` array is first set to `null`, then filtered out/deleted.

jankeromnes replied:
Hmm, you're right, but I see this as an implementation detail of the word `move`. Technically the event representation does get moved to a different collection, even if it leaves behind a temporary empty slot. And that particular behavior is already explained by another comment. But thanks for nitpicking! :smile:

jankeromnes commented:
Woohoo! Thanks a lot! :tada:
